Add function to erase whole nvs partition

schurwanzn-stabl opened the pull request:
Add safe wrapper around `nvs_flash_erase` and `nvs_flash_erase_partition` (https://docs.espressif.com/projects/esp-idf/en/v5.0.2/esp32/api-reference/storage/nvs_flash.html#_CPPv415nvs_flash_erasev) to wipe config partition completely.

schurwanzn-stabl committed:
Add function to erase whole nvs partition

stabl-gjn commented on src/nvs.rs:
```diff
@@ -110,6 +122,19 @@ impl NvsCustom {
     }
+
+    fn erase(partition: &str) -> Result<(), EspError> {
+        let mut registrations = NONDEFAULT_LOCKED.lock();
+        let c_partition = CString::new(partition).unwrap();
```
```suggestion
        let c_partition = CString::new(partition).unwrap();
        let mut registrations = NONDEFAULT_LOCKED.lock();
```

Could switch the lines to reduce time with locked mutex.

schurwanzn-stabl replied:
actually led me to release the lock while erasing since I noticed it might take some time on bigger partitions

stabl-gjn commented on src/nvs.rs:
```diff
@@ -110,6 +122,19 @@ impl NvsCustom {
+        let mut registrations = NONDEFAULT_LOCKED.lock();
+        let c_partition = CString::new(partition).unwrap();
+
+        if registrations.contains(c_partition.as_ref()) {
+            return Err(EspError::from(ESP_ERR_INVALID_STATE).unwrap());
+        }
```
For my understanding: Why is it a invalid state if the registrations contain the partition name?

Maybe also add this as a comment?

stabl-gjn replied:
Comments helped. Thanks :)

schurwanzn-stabl replied:
`esp-idf-svc` doesn't seem to have any proper error reporting for errors originating in it's own code. instead a generic ESP-IDF invalid state error is returned. for what this error means see the now added comments :) good point

stabl-gjn reviewed (approved):
LGTM. Just a really minor thing. Feel free to merge.

schurwanzn-stabl committed:
fix: potential unwanted blocking behaviour
- don't hold on to initialization lock of nvs partition while erasing
- add comments

stabl-gjn commented:
Feel free to merge
